packages/data-context/src/data/ProjectConfigManager.ts

```diff
@@ -0,0 +1,490 @@
+import { CypressError, getError } from '@packages/errors'
+import { IpcHandler, LoadConfigReply, ProjectConfigIpc, SetupNodeEventsReply } from './ProjectConfigIpc'
+import assert from 'assert'
+import type { AllModeOptions, FullConfig, TestingType } from '@packages/types'
+import debugLib from 'debug'
+import path from 'path'
+import _ from 'lodash'
+import chokidar from 'chokidar'
+import { validate as validateConfig, validateNoBreakingConfigLaunchpad, validateNoBreakingConfigRoot, validateNoBreakingTestingTypeConfig } from '@packages/config'
+import { CypressEnv } from './CypressEnv'
+import { autoBindDebug } from '../util/autoBindDebug'
+import type { EventRegistrar } from './EventRegistrar'
+import type { DataContext } from '../DataContext'
+
+const debug = debugLib(`cypress:lifecycle:ProjectConfigManager`)
+
+const UNDEFINED_SERIALIZED = '__cypress_undefined__'
+
+type ProjectConfigManagerOptions = {
+  ctx: DataContext
+  configFile: string | false
+  projectRoot: string
+  handlers: IpcHandler[]
+  hasCypressEnvFile: boolean
+  eventRegistrar: EventRegistrar
+  onError: (cypressError: CypressError, title?: string | undefined) => void
+  onInitialConfigLoaded: (initialConfig: Cypress.ConfigOptions) => void
+  onFinalConfigLoaded: (finalConfig: FullConfig) => Promise<void>
+}
+
+type ConfigManagerState = 'pending' | 'loadingConfig' | 'loadedConfig' | 'loadingNodeEvents' | 'ready' | 'errored'
+
+export class ProjectConfigManager {
+  private _configFilePath: string | undefined
+  private _cachedFullConfig: FullConfig | undefined
+  private _eventsIpc?: ProjectConfigIpc
+  private _pathToWatcherRecord: Record<string, chokidar.FSWatcher> = {}
+  private _watchers = new Set<chokidar.FSWatcher>()
+  private _registeredEventsTarget: TestingType | undefined
+  private _testingType: TestingType | undefined
+  private _state: ConfigManagerState = 'pending'
+  private _loadConfigPromise: Promise<LoadConfigReply> | undefined
+  private _cachedLoadConfig: LoadConfigReply | undefined
+  private _cypressEnv: CypressEnv
+
+  constructor (private options: ProjectConfigManagerOptions) {
+    this._cypressEnv = new CypressEnv({
+      envFilePath: this.envFilePath,
+      validateConfigFile: (filePath, config) => {
+        this.validateConfigFile(filePath, config)
+      },
+    })
+
+    return autoBindDebug(this)
```

Comment:
Also @tgriesser, what does this function do exactly?

If feels like it is helping debug getters, setters and method calls of the current object.
Why don't we add it on every class? what could be the downside?

Reply:
A short comments would be invaluable.

Reply:
Adding here: https://github.com/cypress-io/cypress/pull/20775/commits/1e12aa86126f153a44c0660f36e4bcdc9c6d0867